File: packages/webservice/models/src/main/kotlin/com/yahoo/navi/ws/models/beans/fragments/Request.kt
/**
 * Copyright 2021, Yahoo Holdings Inc.
 * Licensed under the terms of the MIT license. See accompanying LICENSE.md file for terms.
 */
package com.yahoo.navi.ws.models.beans.fragments

import com.fasterxml.jackson.annotation.JsonInclude
import com.yahoo.navi.ws.models.beans.fragments.request.Column
import com.yahoo.navi.ws.models.beans.fragments.request.Filter
import com.yahoo.navi.ws.models.beans.fragments.request.Rollup
import com.yahoo.navi.ws.models.beans.fragments.request.Sort
import com.yahoo.navi.ws.models.types.JsonType
import org.hibernate.annotations.TypeDef

@JsonInclude(JsonInclude.Include.NON_NULL)
@TypeDef(typeClass = JsonType::class, name = "json")
data class Request(
    var filters: List<Filter> = emptyList(),
    var columns: List<Column> = emptyList(),
    var table: String = "",
    var sorts: List<Sort> = emptyList(),
    var rollup: Rollup = Rollup(),
    var limit: Int? = null,
    var dataSource: String? = null,
    var requestVersion: String = "2.0"
)

File: packages/core/addon/serializers/request.js
/**
 * Copyright 2022, Yahoo Holdings Inc.
 * Licensed under the terms of the MIT license. See accompanying LICENSE.md file for terms.
 */
import JSONSerializer from '@ember-data/serializer/json';
import { inject as service } from '@ember/service';
import { canonicalizeColumn } from '@yavin/client/utils/column';
import { assert } from '@ember/debug';

export default class RequestSerializer extends JSONSerializer {
  /**
   * @property {Service}
   */
  @service naviMetadata;

  /**
   * Looks up data source name for a table name
   * @param {String} tableId
   * @returns {string} - data source name
   */
  getTableNamespace(tableId) {
    const tableModel = this.naviMetadata.all('table').find(({ id }) => id === tableId);
    return tableModel.source;
  }

  /**
   * Normalizes the JSON payload returned by the persistence layer
   * @override
   * @param {Model} typeClass
   * @param {Object} request
   * @return {Object} normalized request
   */
  normalize(typeClass, request) {
    const normalized = { ...request };

    //copy source property from request
    ['columns', 'filters', 'sorts'].forEach((attr) =>
      normalized[attr].forEach((fragment) => (fragment.source = normalized.dataSource))
    );

    // add cid to sorts
    const map = new Map();
    normalized.columns?.forEach(({ type, field, parameters, cid }) => {
      const canonicalName = canonicalizeColumn({ field, parameters });
      assert('columns must have cid', cid);
      map.set(`${type}|${canonicalName}`, cid);
    });
    normalized.sorts?.forEach((sort) => {
      const { type, field, parameters } = sort;
      const canonicalName = canonicalizeColumn({ field, parameters });
      const cid = map.get(`${type}|${canonicalName}`);
      assert('sorts must have cid', cid);
      sort.cid = cid;
    });

    return super.normalize(typeClass, normalized);
  }
}
